server/metadata/queue.go

```diff
@@ -0,0 +1,241 @@
+// Copyright 2022-2023 Tigris Data, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package metadata
+
+import (
+	"context"
+	"time"
+
+	jsoniter "github.com/json-iterator/go"
+	"github.com/rs/zerolog/log"
+	"github.com/tigrisdata/tigris/errors"
+	"github.com/tigrisdata/tigris/internal"
+	"github.com/tigrisdata/tigris/keys"
+	"github.com/tigrisdata/tigris/lib/uuid"
+	"github.com/tigrisdata/tigris/server/transaction"
+	"github.com/tigrisdata/tigris/store/kv"
+	ulog "github.com/tigrisdata/tigris/util/log"
+)
+
+const (
+	queueMetaValueVersion int32  = 1
+	itemsSpace            int64  = 1
+	maxId                 string = "$MAX"
+	maxPriority                  = 10
+)
+
+type QueueSubspace struct {
+	metadataSubspace
+}
+
+type QueueItem struct {
+	Id          string `json:"id"`
+	Priority    int64  `json:"priority"`
+	ErrorCount  uint8  `json:"error_count"`
+	LeaseId     string `json:"lease_id,omitempty"`
+	Data        []byte `json:"data"`
+	VestingTime int64  `json:"vesting_time"`
+}
+
+func NewQueueItem(priority int64, data []byte) *QueueItem {
+	return &QueueItem{
+		Id:         uuid.NewUUIDAsString(),
+		Priority:   priority,
+		ErrorCount: 0,
+		Data:       data,
+	}
+}
+
+func newQueueStore(nameRegistry *NameRegistry) *QueueSubspace {
+	return &QueueSubspace{
+		metadataSubspace{
+			SubspaceName: nameRegistry.QueueSubspaceName(),
+			KeyVersion:   []byte{byte(queueMetaValueVersion)},
+		},
+	}
+}
+
+func (q *QueueSubspace) getKey(vestingTime int64, priority int64, id string) keys.Key {
+	if id == maxId {
+		return keys.NewKey(q.SubspaceName, q.KeyVersion, itemsSpace, vestingTime, priority, 0xFF)
+	}
+
+	return keys.NewKey(q.SubspaceName, q.KeyVersion, itemsSpace, vestingTime, priority, id)
+}
+
+func (q *QueueSubspace) Enqueue(ctx context.Context, tx transaction.Tx, item *QueueItem, delay time.Duration) error {
+	if len(item.Id) == 0 {
+		return errors.Internal("cannot enqueue without an id")
+	}
+	item.VestingTime = time.Now().Add(delay).UnixMilli()
```

Comment:
I am assuming the only advantage of vesting time is to reduce the contention between the queue adding items and workers picking it up?

Reply:
Vesting time means an item is never out of the queue and so it is never lost. When a user works on an item, the worker adds the vesting time - it could also be called an expired time at that point. And then if the item is not completed by the time vesting time == current time. Another worker can take over the task.

Reply:
Okay, now I see. So the worst case an item is not picked is the delay set by the worker.